zebra-chain/src/primitives/zcash_history.rs

```diff
@@ -0,0 +1,194 @@
+            new_nodes.push(node);
+        }
+        new_nodes
+    }
+
+    /// Append multiple blocks to the tree.
+    fn append_leaf_iter<'a>(
+        &mut self,
+        vals: impl Iterator<Item = (&'a Block, &'a sapling::tree::Root)>,
+    ) {
+        for (block, root) in vals {
+            self.append_leaf(block, root);
+        }
+    }
+
+    /// Remove the last leaf (block) from the tree.
+    ///
+    /// Returns the number of nodes removed from the tree after the operation.
+    fn truncate_leaf(&mut self) -> u32 {
+        // XXX: handle error
+        self.tree.truncate_leaf().unwrap()
+    }
+
+    /// Return the root hash of the tree, i.e. `hashChainHistoryRoot`.
+    fn hash(&self) -> [u8; 32] {
+        // Both append_leaf() and truncate_leaf() leave a root node, so it should
+        // always exist.
+        self.tree
+            .root_node()
+            .expect("must have root node")
+            .data()
+            .hash()
+    }
+}
+
+/// Convert a Block into a zcash_history::NodeData used in the MMR tree.
+///
+/// `sapling_root` is the root of the Sapling note commitment tree of the block.
+fn convert_block_to_librustzcash_data(
+    block: &Block,
+    network: Network,
+    sapling_root: &sapling::tree::Root,
+) -> zcash_history::NodeData {
+    // XXX: is this expect OK?
+    let height = block
+        .coinbase_height()
+        .expect("block must have coinbase height");
+    // XXX: is this expect OK?
+    let branch_id = ConsensusBranchId::current(network, height).expect("Must have branch ID");
+    let block_hash = block.hash().0;
+    let time: u32 = block
+        .header
+        .time
+        .timestamp()
+        .try_into()
+        .expect("deserialized and generated timestamps are u32 values");
+    let target = block.header.difficulty_threshold.0;
+    let sapling_root: [u8; 32] = (*sapling_root).into();
+    // XXX: is this expect OK?
+    let work = block
+        .header
+        .difficulty_threshold
+        .to_work()
+        .expect("must have work");
+    let work = work.as_u128().to_be_bytes();
+
+    let sapling_tx_count = count_sapling_transactions(block);
+
+    zcash_history::NodeData {
+        consensus_branch_id: branch_id.into(),
+        subtree_commitment: block_hash,
+        start_time: time,
+        end_time: time,
+        start_target: target,
+        end_target: target,
+        start_sapling_root: sapling_root,
+        end_sapling_root: sapling_root,
+        subtree_total_work: (&work[..]).into(),
+        start_height: height.0 as u64,
+        end_height: height.0 as u64,
+        sapling_tx: sapling_tx_count,
+    }
+}
+
+/// Count how many Sapling transactions exist in a block,
+/// i.e. transactions "where either of vSpendsSapling or vOutputsSapling is non-empty"
+/// (https://zips.z.cash/zip-0221#tree-node-specification).
+fn count_sapling_transactions(block: &Block) -> u64 {
+    block
+        .transactions
+        .iter()
+        .map(|tx| -> u64 {
```

Comment:
You might find that type inference does this for you:
```suggestion
        .map(|tx| {
```

Reply:
Neat!